aopy/analysis.py

```diff
@@ -202,6 +239,88 @@ def calc_success_rate(events, start_events=[b"TARGET_ON"], end_events=[b"REWARD"
     success_rate = n_success / n_trials
     return success_rate
 
+'''
+Cell type classification analysis
+'''
+def classify_cells_spike_width(waveform_data, samplerate, std_threshold=3, pca_varthresh=0.75, min_wfs=10):
+    '''
+    Calculates waveform width and classifies units into putative exciatory and inhibitory cell types based on pulse width.
+    Units with lower spike width are considered inhibitory cells (label: 0) and higher spike width are considered excitatory cells (label: 1)
+    The pulse width is defined as the time between the waveform trough to the waveform peak. (trough-to-peak time)
+    Assumes all waveforms are recorded for the same number of time points. 
+
+    This function conducts the following processing steps:
+
+        | **1. ** For each unit, project each waveform into the top PCs. Number of PCs determined by 'pca_varthresh'
+        | **2. ** For each unit, remove outlier spikes. Outlier threhsold determined by 'std_threshold'. If the number of waveforms is less than 'min_wf', no waveforms are removed.
+        | **3. ** For each unit, average remaining waveforms.
+        | **4. ** For each unit, calculate spike width using a local polynomail interpolation.
+        | **5. ** Use a gaussian mixture model to classify all units
+
+    Args:
+        waveform_data (nunit long list of (nt x nwaveforms) arrays): Waveforms of each unit. Each element of the list is a 2D array for each unit. Each 2D array contains the timeseries of all recorded waveforms for a given unit.
+        samplerate (float): sampling rate of the points in each waveform. 
+        std_threshold (float): For outlier removal. The maximum number of standard deviations (in PC space) away from the mean a given waveform is allowed to be. Defaults to 3
+        pca_varthresh (float): Variance threshold for determining the number of dimensions to project spiking data onto. Defaults to 0.75.
+        min_wf_samples (int): Minimum number of waveform samples required to perform outlier detection.
+        nspikesamples (None or int): Optional, defaults to None. Specificy how many waveforms to use in classification. Randomly selects the set amount of waveforms for each unit. If there are more waveforms asked for than recorded spikes all of the recorded spikes will be used. 
```

Comment:
a bit unclear what this parameter does. if it's none, it randomly selects an amount? or if it's an integer then it selects that many randomly and if it's none it takes all of them?

Reply:
The intent was that the function could subselect 'nspikesamples' number of waveforms to average. I think it makes more sense to pull this out of the function and have the user subselect their samples before inputting them into this wrapper.

Reply:
oh wait, i did pull it out and just forgot to remove it from the documentation